gradle-conjure/src/main/java/com/palantir/gradle/conjure/GradleExecUtils.java

```diff
@@ -47,4 +95,52 @@ static void exec(Project project, String failedTo, List<String> unloggedArgs, Li
                     failedTo, combinedArgs, execResult.getExitValue(), output.toString()));
         }
     }
+
+    private static void runJavaCodeInProcess(String failedTo, List<String> combinedArgs, Method mainMethod) {
+        try {
+            String[] args = combinedArgs.toArray(new String[] {});
+            mainMethod.invoke(null, new Object[] {args});
+        } catch (Throwable t) {
+            Throwable rootCause = Throwables.getRootCause(t);
+            if (rootCause instanceof GradleExecStubs.ExitInvoked) {
+                int exitStatus = ((GradleExecStubs.ExitInvoked) rootCause).getExitStatus();
+                if (exitStatus == 0) {
+                    // Exit status zero, we're good to go!
+                    return;
+                }
+                // the error message from a generator attempting to call exit 1 looks pretty gross
+                throw new RuntimeException(String.format(
+                        "Failed to %s. The command '%s' failed with exit code 1. Output above.",
+                        failedTo, combinedArgs));
+            }
+            throw new RuntimeException(
+                    String.format("Failed to %s. The command '%s' failed.", failedTo, combinedArgs), t);
+        }
+    }
+
+    private static Optional<Method> getMainMethod(URLClassLoader classLoader, String mainClassName) {
+        try {
+            ClassFileLocator locator = new ClassFileLocator.ForUrl(classLoader.getURLs());
+            TypePool typePool = TypePool.ClassLoading.of(classLoader);
+            Class<?> mainClass = new ByteBuddy(ClassFileVersion.ofThisVm(ClassFileVersion.JAVA_V8))
+                    .redefine(typePool.describe(mainClassName).resolve(), locator)
+                    .name(mainClassName + "RedefinedForGradleConjure")
+                    .visit(new ForDeclaredMethods()
+                            .invokable(
+                                    ElementMatchers.any(),
+                                    MemberSubstitution.relaxed()
+                                            .method(ElementMatchers.is(System.class.getMethod("exit", int.class)))
+                                            .replaceWith(GradleExecStubs.getStubMethod())))
```

Comment:
So I'm actually a little surprised that this works, because in conjure-java we don't actually have a System.exit() call in the top level mainclass, and instead rely on picocli and all its reflection... https://github.com/palantir/conjure-java/blob/develop/conjure-java/src/main/java/com/palantir/conjure/java/cli/ConjureJavaCli.java#L52-L54

Reply:
how expensive would it be to just visit everything and redefine System.exit everywhere?

Reply:
The conjure cli uses `System.exit(new CommandLine(new ConjureCli()).execute(args));` while the conjure-java cli uses the deprecated `CommandLine.run(new ConjureJavaCli(), args);`. The latter returns on success and throws on error, it never actually calls System.exit.

> how expensive would it be to just visit everything and redefine System.exit everywhere?

Probably relatively expensive, I think we're safe as is since we recommend against calling System.exit from other classes.